### PR TITLE
Add `"depleted"` operator trace handling

### PR DESCRIPTION
**Pull request by szalpal**
Should the `VideoInput` operator be properly working with `ExternalSource` in the same pipeline, DALI Backend needs to properly handle the `"depleted"` operator trace. This trace, defined in DALI, specifies when a given operator needs more data at the input. Refers to https://github.com/NVIDIA/DALI/pull/4794

### CODE REVIEW

**Review by jantonguirao — DISMISSED** [review]

File: src/dali_executor/dali_executor.cc [hunk 1]
```diff
@@ -136,6 +134,32 @@ bool DaliExecutor::IsNoCopy(device_type_t es_device, const IDescr& input) {
           input.buffers[0].device_id == pipeline_.DeviceId());
 }
 
+
+static bool streq(const char * lhs, const char * rhs) {
+  return strcmp(lhs, rhs) == 0;
+}
+
+
+bool DaliExecutor::IsInputConsumed() {
```
Comment (mzient) [inline src/dali_executor/dali_executor.cc:143]: Why not 
```suggestion
bool DaliExecutor::IsInputDepleted() {
```
?

File: src/dali_executor/dali_executor.cc [hunk 2]
```diff
@@ -136,6 +134,32 @@ bool DaliExecutor::IsNoCopy(device_type_t es_device, const IDescr& input) {
...
+  for (auto &name : input_names_) {
+    auto trace = pipeline_.TryGetOperatorTrace(name, "depleted");
+    if (!trace.has_value()) {
+      throw std::runtime_error("DALI Error: 'depleted' trace shall be available for every input operator.");
```
Comment (mzient) [inline src/dali_executor/dali_executor.cc:147]: You're not stating requirements, but rather producing diagnostics. Mentioning the specific name of the operator would be nice, too.
```suggestion
      throw std::logic_error("DALI Error: \"depleted\" trace not found for input \"" + name +"\". "
                             "It must be available for every input operator.");
```

File: src/dali_executor/dali_executor.cc [hunk 3]
```diff
@@ -136,6 +134,32 @@ bool DaliExecutor::IsNoCopy(device_type_t es_device, const IDescr& input) {
...
+    }
+    if (streq(trace->c_str(), "true")) {
+      return true;
+    }
+  }
+  for (auto &name : input_names_) {
+    auto trace = pipeline_.TryGetOperatorTrace(name, "next_output_data_id");
+    if (trace.has_value() && std::stoull(*trace) != request_id_) {
```
Comment (mzient) [inline src/dali_executor/dali_executor.cc:155]: It'd be better to keep the current ID as a string and compare strings.
The code, as is, would say that any non-number matches `request_id_ == 0`.

File: src/dali_executor/dali_executor.cc (hunk 2 above)
Comment (szalpal) [reply src/dali_executor/dali_executor.cc:147]: Done

File: src/dali_executor/request_id.h [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+    return request_id_;
+  }
+
+
+  /**
+   * Postfix increment.
+   */
+  T operator++(int) {
+    T old = request_id_;
+    operator++();
```
Comment (mzient) [inline src/dali_executor/request_id.h:46]: Readability nitpick:
```suggestion
    ++*this;
```

**Review by mzient — APPROVED** [review]

**Review by jantonguirao — APPROVED** [review]

**Pull request merged**